BAD SAD Error when a mismatching constraint type is provided for tables returned from SQL Connector

**Description:**
```bash
[2018-12-07 12:18:02,170] ERROR {org.ballerinalang.launcher.Main} - org.ballerinalang.util.exceptions.BallerinaException: input value John1 cannot be cast to integer 
java.lang.RuntimeException: org.ballerinalang.util.exceptions.BallerinaException: input value John1 cannot be cast to integer
	at org.ballerinalang.bre.bvm.BVMScheduler.execute(BVMScheduler.java:77)
	at org.ballerinalang.bre.bvm.BVMExecutor.execute(BVMExecutor.java:180)
	at org.ballerinalang.bre.bvm.BVMExecutor.executeEntryFunction(BVMExecutor.java:92)
	at org.ballerinalang.BLangProgramRunner.runEntryFunc(BLangProgramRunner.java:74)
	at org.ballerinalang.launcher.LauncherUtils.runMain(LauncherUtils.java:168)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:160)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:299)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:67)
Caused by: org.ballerinalang.util.exceptions.BallerinaException: input value John1 cannot be cast to integer
	at org.ballerinalang.model.values.BString.intValue(BString.java:49)
	at org.ballerinalang.bre.bvm.BVM.execTypeCastOpcodes(BVM.java:2572)
	at org.ballerinalang.bre.bvm.BVM.execute(BVM.java:604)
	at org.ballerinalang.bre.bvm.BVMScheduler.execute(BVMScheduler.java:72)
	... 8 more
```

--- a/stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java
+++ b/stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java
@@ -140,6 +140,12 @@ public class SQLDataIterator extends TableIterator {
         int sqlType = -1;
         try {
             BField[] structFields = this.type.getFields().values().toArray(new BField[0]);
+            if (columnDefs.size() != structFields.length) {
+                throw new BallerinaException(
+                        "Number of fields in the constraint type is " + (structFields.length > columnDefs.size() ?
+                                "greater" :
+                                "lower") + " than column count of the result set");
+            }
             for (ColumnDefinition columnDef : columnDefs) {
                 if (columnDef instanceof SQLColumnDefinition) {
                     SQLColumnDefinition def = (SQLColumnDefinition) columnDef;
@@ -442,8 +448,8 @@ public class SQLDataIterator extends TableIterator {
                 handleUnAssignableUnionTypeAssignment();
             }
         } else if (fieldTypeTag == TypeTags.RECORD_TYPE_TAG) {
-            bStruct.put(fieldName,
-                    createUserDefinedType(structData, (BRecordType) fieldType));
+            validateAndSetRefRecordField(bStruct, fieldName, TypeTags.RECORD_TYPE_TAG, fieldTypeTag,
+                    createUserDefinedType(structData, (BRecordType) fieldType), MISMATCHING_FIELD_ASSIGNMENT);
         } else {
             handleMismatchingFieldAssignment();
         }
@@ -461,7 +467,8 @@ public class SQLDataIterator extends TableIterator {
             if (isOriginalValueNull) {
                 handleNilToNonNillableFieldAssignment();
             } else {
-                bStruct.put(fieldName, new BBoolean(boolValue));
+                validateAndSetRefRecordField(bStruct, fieldName, TypeTags.BOOLEAN_TAG, fieldTypeTag,
+                        new BBoolean(boolValue), MISMATCHING_FIELD_ASSIGNMENT);
             }
         }
     }
@@ -493,7 +500,16 @@ public class SQLDataIterator extends TableIterator {
             }
         } else {
             if (bytes != null) {
-                bStruct.put(fieldName, new BValueArray(bytes));
+                if (TypeTags.ARRAY_TAG == fieldTypeTag) {
+                    int elementTypeTag = ((BArrayType) fieldType).getElementType().getTag();
+                    if (elementTypeTag == TypeTags.BYTE_TAG) {
+                        bStruct.put(fieldName, new BValueArray(bytes));
+                    } else {
+                        throw new BallerinaException(MISMATCHING_FIELD_ASSIGNMENT);
+                    }
+                } else {
+                    throw new BallerinaException(MISMATCHING_FIELD_ASSIGNMENT);
+                }
             } else {
                 handleNilToNonNillableFieldAssignment();
             }
@@ -509,7 +525,8 @@ public class SQLDataIterator extends TableIterator {
                     retrieveNonNilTypeTag(fieldType), refValue, UNASSIGNABLE_UNIONTYPE_EXCEPTION);
         } else {
             if (stringValue != null) {
-                bStruct.put(fieldName, new BString(stringValue));
+                validateAndSetRefRecordField(bStruct, fieldName, TypeTags.STRING_TAG, fieldTypeTag,
+                        new BString(stringValue), MISMATCHING_FIELD_ASSIGNMENT);
             } else {
                 handleNilToNonNillableFieldAssignment();
             }
@@ -573,7 +590,8 @@ public class SQLDataIterator extends TableIterator {
             if (isOriginalValueNull) {
                 handleNilToNonNillableFieldAssignment();
             } else {
-                bStruct.put(fieldName, new BInteger(longValue));
+                validateAndSetRefRecordField(bStruct, fieldName, TypeTags.INT_TAG, fieldTypeTag,
+                        new BInteger(longValue), MISMATCHING_FIELD_ASSIGNMENT);
             }
         }
     }
@@ -590,7 +608,8 @@ public class SQLDataIterator extends TableIterator {
             if (isOriginalValueNull) {
                 handleNilToNonNillableFieldAssignment();
             } else {
-                bStruct.put(fieldName, new BFloat(fValue));
+                validateAndSetRefRecordField(bStruct, fieldName, TypeTags.FLOAT_TAG, fieldTypeTag,
+                        new BFloat(fValue), MISMATCHING_FIELD_ASSIGNMENT);
             }
         }
     }

--- a/stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java
+++ b/stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java
@@ -143,8 +143,7 @@ public class SQLDataIterator extends TableIterator {
             if (columnDefs.size() != structFields.length) {
                 throw new BallerinaException(
                         "Number of fields in the constraint type is " + (structFields.length > columnDefs.size() ?
-                                "greater" :
-                                "lower") + " than column count of the result set");
+                                "greater" : "lower") + " than column count of the result set");
             }
             for (ColumnDefinition columnDef : columnDefs) {
                 if (columnDef instanceof SQLColumnDefinition) {

--- a/tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java
+++ b/tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java
@@ -57,6 +57,7 @@ import java.util.Calendar;
 public class TableTest {
 
     private CompileResult result;
+    private CompileResult resultNegative;
     private CompileResult nillableMappingNegativeResult;
     private CompileResult nillableMappingResult;
     private static final String DB_NAME_H2 = "TEST_DATA_TABLE_H2";
@@ -77,6 +78,7 @@ public class TableTest {
                 SQLDBUtils.DB_DIRECTORY, DB_NAME_H2);
 
         result = BCompileUtil.compile("test-src/types/table/table_type.bal");
+        resultNegative = BCompileUtil.compile("test-src/types/table/table_type_negative.bal");
         nillableMappingNegativeResult = BCompileUtil
                 .compile("test-src/types/table/table_nillable_mapping_negative.bal");
         nillableMappingResult = BCompileUtil.compile("test-src/types/table/table_nillable_mapping.bal");
@@ -1257,5 +1259,67 @@ public class TableTest {
           expectedExceptionsMessageRegExp = ".*Table query over a cursor table not supported.*")
     public void testJoinQueryWithCursorTable() {
         BRunUtil.invoke(result, "testJoinQueryWithCursorTable");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderInt() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderInt");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderString() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderString");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderBoolean() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderBoolean");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderFloat() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderFloat");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderDouble() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderDouble");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderLong() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderLong");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testWrongOrderBlob() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderBlobWrongOrder");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Trying to assign to a mismatching type.*")
+    public void testCorrectOrderWrongTypeBlob() {
+        BRunUtil.invoke(resultNegative, "testWrongOrderBlobCorrectOrderWrongType");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Number of fields in the constraint type is greater than column count "
+                  + "of the result set.*")
+    public void testGreaterNoOfParams() {
+        BRunUtil.invoke(resultNegative, "testGreaterNoOfParams");
+    }
+
+    @Test(expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*Number of fields in the constraint type is lower than column count of "
+                  + "the result set.*")
+    public void testLowerNoOfParams() {
+        BRunUtil.invoke(resultNegative, "testLowerNoOfParams");
     }
 }

--- a/tests/ballerina-unit-test/src/test/resources/test-src/jdbc/sql_actions_test.bal
+++ b/tests/ballerina-unit-test/src/test/resources/test-src/jdbc/sql_actions_test.bal
@@ -22,7 +22,6 @@ type ResultBlob record {
 
 type ResultRowIDBlob record {
     int row_id;
-    byte[] BLOB_TYPE;
 };
 
 type ResultDataType record {

--- a/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal
+++ b/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal
@@ -53,7 +53,7 @@ type ResultDatesWithNillableIntType record {
     int? DATETIME_TYPE;
 };
 
-type NillableDataTypes record {
+type NillableDataTypesAll record {
     int? int_type;
     int? long_type;
     float? float_type;
@@ -71,6 +71,22 @@ type NillableDataTypes record {
     time:Time? time_type;
     time:Time? datetime_type;
     time:Time? timestamp_type;
+};
+
+type NillableDataTypes record {
+    int? int_type;
+    int? long_type;
+    float? float_type;
+    float? double_type;
+    boolean? boolean_type;
+    string? string_type;
+    float? numeric_type;
+    float? decimal_type;
+    float? real_type;
+    int? tinyint_type;
+    int? smallint_type;
+    string? clob_type;
+    byte[]? binary_type;
 };
 
 type NillableBlob record {
@@ -164,7 +180,7 @@ function testMappingToNillableTypeFieldsBlob() returns (byte[]?) {
     byte[]? blob_type = ();
     var dt = testDB->select("SELECT blob_type from DataTypeTableNillableBlob where
     row_id=3", NillableBlob);
-    if (dt is table<NillableDataTypes>) {
+    if (dt is table<NillableBlob>) {
         while (dt.hasNext()) {
             var rs = dt.getNext();
             if (rs is NillableBlob) {
@@ -334,7 +350,7 @@ function testMappingNullToNillableTypes() returns (int?, int?, float?,
     var dt = testDB->select("SELECT int_type, long_type, float_type, double_type,
     boolean_type, string_type, numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type,
     binary_type, date_type, time_type, datetime_type, timestamp_type from DataTypeTableNillable where
-    row_id=2", NillableDataTypes);
+    row_id=2", NillableDataTypesAll);
 
     int? int_type = ();
     int? long_type = ();
@@ -354,10 +370,10 @@ function testMappingNullToNillableTypes() returns (int?, int?, float?,
     time:Time? datetime_type = ();
     time:Time? timestamp_type = ();
 
-    if (dt is table<NillableDataTypes>) {
+    if (dt is table<NillableDataTypesAll>) {
         while (dt.hasNext()) {
             var rs = dt.getNext();
-            if (rs is NillableDataTypes) {
+            if (rs is NillableDataTypesAll) {
                 int_type = rs.int_type;
                 long_type = rs.long_type;
                 float_type = rs.float_type;

--- a/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_type_negative.bal
+++ b/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_type_negative.bal
@@ -1,0 +1,159 @@
+// Copyright (c) 2018 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/sql;
+import ballerina/h2;
+import ballerina/io;
+
+type Result record {
+    string STRING_TYPE;
+    int INT_TYPE;
+    int LONG_TYPE;
+    float FLOAT_TYPE;
+    float DOUBLE_TYPE;
+    boolean BOOLEAN_TYPE;
+};
+
+type ResultBlobWrongOrder record {
+    int INT_TYPE;
+    byte[] BLOB_TYPE;
+};
+
+type ResultBlobCorrectOrderWrongType record {
+    int[] BLOB_TYPE;
+    int INT_TYPE;
+};
+
+function testWrongOrderInt() {
+    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderString() {
+    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderBoolean() {
+    testWrongOrder("SELECT boolean_type, long_type, float_type, string_type, int_type, double_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderDouble() {
+    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderFloat() {
+    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderLong() {
+    testWrongOrder("SELECT boolean_type, string_type, float_type, long_type, int_type, double_type from DataTable WHERE row_id = 1");
+}
+
+function testWrongOrderBlobWrongOrder() {
+    h2:Client testDB = new({
+            path: "./target/tempdb/",
+            name: "TEST_DATA_TABLE_H2",
+            username: "SA",
+            password: "",
+            poolOptions: { maximumPoolSize: 1 }
+        });
+
+    var selectRet = testDB->select("SELECT blob_type, row_id from ComplexTypes WHERE row_id = 1", ResultBlobWrongOrder);
+
+    if (selectRet is table<ResultBlobWrongOrder>) {
+        while (selectRet.hasNext()) {
+            var rs = selectRet.getNext();
+        }
+    }
+    testDB.stop();
+}
+
+function testWrongOrderBlobCorrectOrderWrongType() {
+    h2:Client testDB = new({
+            path: "./target/tempdb/",
+            name: "TEST_DATA_TABLE_H2",
+            username: "SA",
+            password: "",
+            poolOptions: { maximumPoolSize: 1 }
+        });
+
+    var selectRet = testDB->select("SELECT blob_type, row_id from ComplexTypes WHERE row_id = 1",
+        ResultBlobCorrectOrderWrongType);
+
+    if (selectRet is table<ResultBlobCorrectOrderWrongType>) {
+        while (selectRet.hasNext()) {
+            var rs = selectRet.getNext();
+        }
+    }
+    testDB.stop();
+}
+
+function testGreaterNoOfParams() {
+    h2:Client testDB = new({
+            path: "./target/tempdb/",
+            name: "TEST_DATA_TABLE_H2",
+            username: "SA",
+            password: "",
+            poolOptions: { maximumPoolSize: 1 }
+        });
+
+    var selectRet = testDB->select("SELECT boolean_type from DataTable WHERE row_id = 1", Result);
+
+    if (selectRet is table<Result>) {
+        while (selectRet.hasNext()) {
+            var rs = selectRet.getNext();
+        }
+    }
+    testDB.stop();
+}
+
+function testLowerNoOfParams() {
+    h2:Client testDB = new({
+            path: "./target/tempdb/",
+            name: "TEST_DATA_TABLE_H2",
+            username: "SA",
+            password: "",
+            poolOptions: { maximumPoolSize: 1 }
+        });
+
+    var selectRet = testDB->select("SELECT boolean_type, boolean_type, string_type, float_type, long_type, int_type,
+        double_type from DataTable WHERE row_id = 1", Result);
+
+    if (selectRet is table<Result>) {
+        while (selectRet.hasNext()) {
+            var rs = selectRet.getNext();
+        }
+    }
+    testDB.stop();
+}
+
+function testWrongOrder(string queryStr) {
+    h2:Client testDB = new({
+            path: "./target/tempdb/",
+            name: "TEST_DATA_TABLE_H2",
+            username: "SA",
+            password: "",
+            poolOptions: { maximumPoolSize: 1 }
+        });
+
+    var selectRet = testDB->select(queryStr, Result);
+
+    if (selectRet is table<Result>) {
+        while (selectRet.hasNext()) {
+            var rs = selectRet.getNext();
+        }
+    }
+    testDB.stop();
+}

--- a/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_type_negative.bal
+++ b/tests/ballerina-unit-test/src/test/resources/test-src/types/table/table_type_negative.bal
@@ -14,9 +14,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
-import ballerina/sql;
 import ballerina/h2;
-import ballerina/io;
 
 type Result record {
     string STRING_TYPE;
@@ -38,27 +36,33 @@ type ResultBlobCorrectOrderWrongType record {
 };
 
 function testWrongOrderInt() {
-    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderString() {
-    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT int_type, long_type, float_type, boolean_type, string_type, double_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderBoolean() {
-    testWrongOrder("SELECT boolean_type, long_type, float_type, string_type, int_type, double_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT boolean_type, long_type, float_type, string_type, int_type, double_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderDouble() {
-    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderFloat() {
-    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT double_type, long_type, float_type, string_type, int_type, boolean_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderLong() {
-    testWrongOrder("SELECT boolean_type, string_type, float_type, long_type, int_type, double_type from DataTable WHERE row_id = 1");
+    testWrongOrder("SELECT boolean_type, string_type, float_type, long_type, int_type, double_type from DataTable
+    WHERE row_id = 1");
 }
 
 function testWrongOrderBlobWrongOrder() {

